proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +251,224 @@ protocol TargetBuildContext {
+    /// inputs.
+    ///
+    /// Because prebuild commands are run on every build, they are can have
+    /// significant performance impact and should only be used when there is
+    /// no way to know the names of the outputs before the command is run.
+    ///
+    /// The `outputDirectory` parameter is the path of a directory into which
+    /// the command will write its output files. Any file in that directory
+    /// will be interpreted according to same build rules as for sources.
+    func addPrebuildCommand(
+        /// An arbitrary string to show in build logs and other status areas.
+        displayName: String,
+        /// The executable to be invoked; should be a tool looked up using
+        /// `tool(named:)`, which may reference either a binary-provided tool
+        /// or a source-built tool.
+        executable: Path,
+        /// Arguments to be passed to the tool. Any paths should be based on
+        /// the paths provided in the target build context.
+        arguments: [String],
+        /// Optional initial working directory of the command.
+        workingDirectory: Path? = nil,
+        /// Any custom settings for the process environment.
+        environment: [String: String] = [:],
+        /// A directory into which the command can write output files that
+        /// should be processed further.
+        outputDirectory: Path
+    )
 }
 
 /// Emits errors, warnings, and remarks to be shown as a result of running
 /// the plugin. If any errors are emitted, the plugin is considered to have
 /// have failed, which will be reported to users during the build.
 protocol DiagnosticsEmitter {
-    func emit(error message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(warning message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(remark message: String, file: FilePath? = nil, line: Int? = nil)
+    func emit(error message: String, file: Path? = nil, line: Int? = nil)
+    func emit(warning message: String, file: Path? = nil, line: Int? = nil)
+    func emit(remark message: String, file: Path? = nil, line: Int? = nil)
 }
 
-/// In addition, the API of `FilePath` from SwiftSystem will be available.
+/// Provides information about a list of files. The order is not defined
+/// but is guaranteed to be stable. This allows the implementation to be
+/// more efficient than a static file list.
+protocol FileList: Sequence {
+    func makeIterator() -> FileListIterator
+}
+struct FileListIterator: IteratorProtocol {
+    mutating func next() -> FileInfo?
+}
+
+/// Provides information about a single file in a FileList.
+protocol FileInfo {
+    var path: Path { get }
+    var type: FileType { get }
+}
+
+/// Provides information about a file type. This is expressed as a pro-
+/// tocol so that it can be extended without affecting existing plugins.
+/// It provides additional information about the inputs to the plugin.
+protocol FileType {
```

Comment:
would have been nice to use an enum here, I assume you chose a protocol since adding cases to enum breaks API compatibility?

Reply:
Yes, I started off with this as an enum, but the API compatibility seemed like a deal breaker.  I am not nearly as good with Swift in this area — is there some way of having an open enum?  I couldn't find a way, and so went for more of a traits-based approach where the code can just ask for the traits that are exposed by the types.

Reply:
Since I was wrong about not being able to put `@available()` on an individual enum case (you can), I will change this back to an enum.